onnx/common/path.h

```diff
@@ -8,17 +8,67 @@
 #pragma once
 
 #include <string>
+#ifdef _WIN32
+#include <windows.h>
+#endif
 
 namespace ONNX_NAMESPACE {
 
 #ifdef _WIN32
 const std::string k_preferred_path_separator = "\\";
+const std::wstring w_k_preferred_path_separator = L"\\";
 #else // POSIX
 const std::string k_preferred_path_separator = "/";
 #endif
 
-std::string path_join(const std::string& origin, const std::string& append);
-void normalize_separator(std::string& path);
-std::string clean_relative_path(const std::string& path);
+template <typename STRING>
+STRING path_join(const STRING& origin, const STRING& append);
+
+template <typename CHAR>
+bool is_path_separator(CHAR c) {
+  // Windows accept / as path separator.
+  if (k_preferred_path_separator == "\\") {
+    return c == '\\' || c == '/';
+  }
+
+  return c == k_preferred_path_separator[0];
+}
+
+// To always use backward slash as path separator on Windows
+template <typename STRING>
+void normalize_separator(STRING& path) {
+  char preferred_sep = k_preferred_path_separator[0];
+  if (preferred_sep == '/') {
+    // Do nothing on linux.
+    return;
+  }
+
+  for (size_t i = 0; i < path.size(); i++) {
+    if (is_path_separator(path[i]) && path[i] != preferred_sep) {
+      path[i] = preferred_sep;
+    }
+  }
+}
+
+// Clean up relative path when there is ".." in the path, e.g.: a/b/../c -> a/c
+// std::string or std::wstring
+template <typename STRING>
+STRING clean_relative_path(const STRING& path);
+
+// std::string in template cannot be recognized as char[]
+// Therefore explicitly define char[] to handle char[] input
+template <typename CHAR, typename std::size_t N>
+std::basic_string<CHAR> clean_relative_path(const CHAR (&path)[N]) {
```

Comment:
What is `const CHAR (&path)[N]` ? I'm not familiar with this grammar .

Reply:
This is the way to create template for array of T I believe: https://stackoverflow.com/questions/33234979/how-to-write-a-template-function-that-takes-an-array-and-an-int-specifying-array

Reply:
Previously I tried C-string pointer and it didn't work. It seems required to be defined like it for using template with char[] string (i.e., "this is a string"). Still, now Windows and POSIX have different implementation so this PR doesn't need template anymore.